Test codeclimate

MPParsley opened the pull request:
You can ignore this PR. It's a test to demonstrate CodeClimate.

MPParsley committed:
Test codeclimate

MPParsley committed:
Create .codeclimate.yml

MPParsley commented:
I enabled code climate and configured some defaults.

Some of these checks may be somewhat rigorous (e.g. the limit for the amount of lines in a method is now 50). This is a good default but obviously this doesn't work for Drupal Form classes. Note that it is possible to dismiss suggestions.

If we plan to enable branch protection with code climate, we need to make sure all contributors have access to dismiss these suggestions.

@amitaibu, can you check if you can [dismiss suggestions](https://codeclimate.com/github/Gizra/og/pull/506)?

From within the PR:
<img width="1095" alt="dismiss_check" src="https://user-images.githubusercontent.com/1823998/57629557-438a5d00-759c-11e9-8d2d-351ba2ba410c.png">

On codeclimate itself:
<img width="812" alt="dismiss_codeclimate" src="https://user-images.githubusercontent.com/1823998/57629664-73396500-759c-11e9-87d3-da397f2ea7e6.png">

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+    path: ".csslintrc"
+  - url: "https://raw.githubusercontent.com/digipolisgent/ci_config-files-and-scripts/master/php/drupal8/.eslintrc.json"
+    path: ".eslintrc.json"
+  - url: "https://raw.githubusercontent.com/digipolisgent/ci_config-files-and-scripts/master/php/drupal8/.eslintignore"
```
```suggestion
  - url: "https://git.drupalcode.org/project/drupal/raw/8.7.x/core/.eslintignore"
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+    path: ".phpmd.xml"
+  - url: "https://raw.githubusercontent.com/digipolisgent/ci_config-files-and-scripts/master/php/drupal8/.csslintrc"
+    path: ".csslintrc"
+  - url: "https://raw.githubusercontent.com/digipolisgent/ci_config-files-and-scripts/master/php/drupal8/.eslintrc.json"
```
```suggestion
  - url: "https://git.drupalcode.org/project/drupal/raw/8.7.x/core/.eslintrc.json"
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+  fetch:
+  - url: "https://raw.githubusercontent.com/digipolisgent/ci_config-files-and-scripts/master/php/drupal8/.phpmd.xml"
+    path: ".phpmd.xml"
+  - url: "https://raw.githubusercontent.com/digipolisgent/ci_config-files-and-scripts/master/php/drupal8/.csslintrc"
```
```suggestion
  - url: "https://git.drupalcode.org/project/drupal/raw/8.7.x/.csslintrc"
```

MPParsley committed:
Merge pull request #507 from Gizra/8.x-1.x

Merge develop

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+      threshold: 4
+  file-lines:
+    config:
+      threshold: 250
```
```suggestion
      threshold: 1000
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+    enabled: false
+  complex-logic:
+    config:
+      threshold: 4
```
```suggestion
      threshold: 16
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+      threshold: 250
+  method-complexity:
+    config:
+      threshold: 5
```
```suggestion
      threshold: 20
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+      threshold: 5
+  method-count:
+    config:
+      threshold: 20
```
```suggestion
      threshold: 80
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+      threshold: 20
+  method-lines:
+    config:
+      threshold: 50
```
```suggestion
      threshold: 200
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+      threshold: 50
+  nested-control-flow:
+    config:
+      threshold: 4
```
```suggestion
      threshold: 16
```

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,84 @@
+      threshold: 4
+  return-statements:
+    config:
+      threshold: 4
```
```suggestion
      threshold: 16
```

MPParsley committed:
Alleviate thresholds x4

MPParsley committed:
Use .csslintrc from Drupal core

MPParsley committed:
Use .eslintrc.json from Drupal core

MPParsley committed:
Use .eslintignore from Drupal core

MPParsley committed:
Add codeclimate to travis

MPParsley commented on src/GroupTypeManager.php:
```diff
@@ -8,8 +8,7 @@
 use Drupal\Core\Routing\RouteBuilderInterface;
 use Drupal\Core\Cache\CacheBackendInterface;
 use Drupal\og\Event\GroupCreationEvent;
```
```suggestion
use Drupal\og\Event\GroupCreationEvent;
use Drupal\og\Event\GroupCreationEventInterface;
use Symfony\Component\EventDispatcher\EventDispatcherInterface;
```

MPParsley committed:
Revert test

pfrenssen commented:
I'm starting to see the value in this :smile:

Can't you disable the checks on cyclomatic complexity, and the ones on method length and file length? Just increasing the limit doesn't make sense I think.

MPParsley commented on src/GroupTypeManager.php:
```diff
@@ -10,6 +10,7 @@
 use Drupal\og\Event\GroupCreationEvent;
 use Drupal\og\Event\GroupCreationEventInterface;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Drupal\node\Entity\Node;
```
```suggestion
```

MPParsley committed:
Cleanup test

MPParsley committed:
Disable some plugins

MPParsley commented:
Yes, we should definitely finetune this.

- disabled some of the "dumb" checks (line count etc);
- added some linters from Drupal core;
- we can configure test coverage to measure and enforce it

I hope the infrastructure team will provide something similar on Gitlab but in the meantime we can get some added value from this tool.

Note: I'm not 100% sure but I think that we must merge the .codeclimate.yml file into develop before it takes place.

MPParsley commented on .codeclimate.yml:
```diff
@@ -0,0 +1,78 @@
+  phpcodesniffer:
+    enabled: true
+    config:
+      standard: "Drupal"
```
Let's make it really interesting:
```suggestion
      standard: "Drupal,DrupalPractice"
```

This detects issues like:
- Global constants should not be used, move it to a class or interface
- t() calls should be avoided in classes, use dependency injection and $this->t() instead
- User::load calls should be avoided in classes, use dependency injection instead
- ...

MPParsley committed:
Add DrupalPractice phpcs standard

pfrenssen commented:
Wow now it is full of actionable suggestions!! This is very nice!

We could make a whole bunch of novice issues out of this for using at the Drupal Dev Days to do sprint mentoring!

MPParsley committed:
Added coverage to phpunit

MPParsley committed:
Keep a copy of xdebug

MPParsley committed:
Re-enable xdebug to support PHPUnit code coverage.

MPParsley committed:
Re-enable xdebug to support PHPUnit code coverage.

MPParsley committed:
Update .travis.yml

MPParsley committed:
Added documentation for Xdebug

- Xdebug is needed for code coverage
- Removed deprecated comment (in the past Xdebug wasn't supported for PHP 7)

MPParsley commented:
Phpunit requires Xdebug and Xdebug makes phpunit a lot slower. So now the tests failed after 50 minutes.

```
Testing /home/travis/build/Gizra/drupal/modules/og/tests
...................
The job exceeded the maximum time limit for jobs, and has been terminated.
```

We'll have to do without code coverage for now. 😞

MPParsley commented on .travis.yml:
```diff
@@ -76,4 +75,15 @@ before_script:
+  - ./cc-test-reporter before-build
+
+  # Re-enable xdebug to support PHPUnit code coverage.
+  - if [ "$TRAVIS_PULL_REQUEST" == "false" ] && [ "${TEST_SUITE}" != "PHP_CodeSniffer" ]; then phpenv config-add /tmp/xdebug.ini; fi
```
```suggestion
  # - if [ "$TRAVIS_PULL_REQUEST" == "false" ] && [ "${TEST_SUITE}" != "PHP_CodeSniffer" ]; then phpenv config-add /tmp/xdebug.ini; fi
```

MPParsley committed:
Disable xdebug

MPParsley committed:
Added configuration for PHP Mess Detector

Source: https://raw.githubusercontent.com/digipolisgent/ci_config-files-and-scripts/master/php/drupal8/.phpmd.xml

MPParsley committed:
Update .codeclimate.yml

pfrenssen commented:
Can we disable PHP_CodeSniffer? This is not needed since this is already being checked, and it is throwing errors: https://codeclimate.com/github/Gizra/og/pull/509

pfrenssen commented:
I wouldn't run the tests through CodeClimate either. These are already running very reliably.

Rule #1: if it is not broken, don't fix it.

pfrenssen commented:
It's still reporting a lot of non-issues, and it seems the test code is being ignored?

I would disable PHP Mess Detector I have never known this to give useful feedback.

pfrenssen commented:
Also, would it be possible to experiment with this on a fork instead of on the main repository? Currently all the ongoing PRs are failing because of this. If we can get it to a workable state on a fork it should be easy to move it over to the main repo after. And then at least we can continue working without interruption.

MPParsley committed:
Disable PHP Mess Detection

MPParsley committed:
Delete .phpmd.xml

MPParsley commented:
Some overdue feedback;

> Can we disable PHP_CodeSniffer? This is not needed since this is already being checked, and it is throwing errors: https://codeclimate.com/github/Gizra/og/pull/509

Preferably not. Having feedback inline in the code on GitHub is an added value over the Travis report.

> I wouldn't run the tests through CodeClimate either. These are already running very reliably.

Note that we're never running tests through CodeClimate. We merely send an xml report to CodeClimate.

> Rule #1: if it is not broken, don't fix it.

I'll comment on this quote with [a few other quotes](https://www.brainyquote.com/topics/innovation) ;-)

This feature would enable us to:
- enforce partial (percentage) or even full code coverage by tests
- enforce same percentage or more code coverage between PR's
- visible reports and statistics on the progress we're making

> It's still reporting a lot of non-issues, and it seems the test code is being ignored?

In order for phpunit to measure test coverage, Xdebug is needed. Unfortunately I ran into this issue in .travis.yml and it takes more than 50 minutes to run all tests on Travis (might be in part due to the amount of Kernel tests but this would need further investigation):
```
  # Remove Xdebug as we don't need it and it causes "PHP Fatal error: Maximum
  # function nesting level of '256' reached."
  # We also don't care if that file exists or not on PHP 7.
```

For now I disabled this feature as you requested.

> I would disable PHP Mess Detector I have never known this to give useful feedback.

I disabled this feature as you requested but I'm convinced that all this feedback is welcome:

```
PHPMD looks for several potential problems within PHP source code, like possible bugs, suboptimal code, overcomplicated expressions, and unused parameters, methods, or properties.
```

If there's still feedback that we consider useless, CodeClimate allows you to [simply ignore this](https://codeclimate.com/changelog/5670481aac74390098001828/).

> Also, would it be possible to experiment with this on a fork instead of on the main repository? Currently all the ongoing PRs are failing because of this. If we can get it to a workable state on a fork it should be easy to move it over to the main repo after. And then at least we can continue working without interruption.

I had set it up on my fork to but the errors you see in the other PR's are based on the default configuration of CodeClimate. Until we add the .codeclimate.yml configuration in #505 the default configuration will be used (which has a wrong CS ruleset that report 4 spaces instead of 2 for example).

If you like **I can disable the integration completely** or **we could commit the .codeclimate.yml configuration from #505** which now only reports 63 issues (btw, which is awesome < %5 technical debt). I marked them all as approved so they won't show up in new PR's. We could then resolve them in follow up issues or choose to ignore them and benefit from the feedback from CodeClimate in new PR's.

wdyt?

cc @pfrenssen, @amitaibu, @AronNovak
